### PR TITLE
[FLINK-23708][FLINK-23698][task] Re emit and pass watermarks in finished on restore tasks 

### PR DESCRIPTION
**Pull request by pnowojski**
This PR provides a bug fixes for two bugs: FLINK-23708 and FLINK-23698

## Verifying this change

Add a various unit tests for missing test coverage

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / no ****/ don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskFinalCheckpointsTest.java [hunk 1]
```diff
@@ -712,18 +714,24 @@ public void testOperatorSkipLifeCycleIfFinishedOnRestore() throws Exception {
         try (StreamTaskMailboxTestHarness<String> harness =
                 new StreamTaskMailboxTestHarnessBuilder<>(
                                 OneInputStreamTask::new, BasicTypeInfo.STRING_TYPE_INFO)
+                        .setCollectNetworkEvents()
                         .addInput(BasicTypeInfo.STRING_TYPE_INFO, 3)
+                        .setCollectNetworkEvents()
```
Comment (gaoyunhaii) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskFinalCheckpointsTest.java:719]: The setting is repeated.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskChainedSourcesCheckpointingTest.java [hunk 2]
```diff
@@ -507,9 +507,6 @@ public void notifyEndOfData() throws IOException {
                     sourceReader.getLifeCycleMonitor().assertCallTimes(0, LifeCyclePhase.values());
                 } else if (wrapper.getStreamOperator()
```
Comment (gaoyunhaii) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskChainedSourcesCheckpointingTest.java:508]: Remove the empty branch ?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskChainedSourcesCheckpointingTest.java [hunk 3]
```diff
@@ -490,12 +494,8 @@ public void notifyEndOfData() throws IOException {
                         .finishForSingletonOperatorChain(StringSerializer.INSTANCE)
                         .build()) {
 
-            testHarness.processElement(Watermark.MAX_WATERMARK, 0, 0);
```
Comment (gaoyunhaii) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskChainedSourcesCheckpointingTest.java:493]: We may keep emitting the MAX_WATERMARK and check the MAX_WATERMARK is passed ?

**Review by gaoyunhaii — COMMENTED** [review]
Very thanks @pnowojski  opening the PR. The PR in general looks good to me:
1. I agree we could extract a specialized `FinishedOperatorChain` to skip the execution as a whole to avoid missing required modification.
2. Currently we do not remove the calls to `getMetricGroup`, but it should not cause problem for now.
3. Bypass the watermark if finished on restore for all the types of the tasks. 

I left some comments on some minor issues.

**Review by pnowojski — COMMENTED** [review]
Thanks @gaoyunhaii  for the review. Please check my comments/fixup commits.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 4]
```diff
@@ -609,16 +586,13 @@ private boolean takeSnapshotSync(
...
-                operatorChain.getAllOperators(true)) {
-            if (!enableCheckpointAfterTasksFinished && operatorWrapper.isClosed()) {
-                env.declineCheckpoint(
-                        checkpointMetaData.getCheckpointId(),
-                        new CheckpointException(
-                                "Task Name" + taskName,
-                                CheckpointFailureReason.CHECKPOINT_DECLINED_TASK_CLOSING));
-                return false;
-            }
+        if (!enableCheckpointAfterTasksFinished && operatorChain.isClosed()) {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java]: I've added a separate hotfix commit to change this into `checkState(!operatorChain.isClosed())`

**Review by gaoyunhaii — COMMENTED** [review]
Very thanks Piotr @pnowojski for the update! I only have one minor issue~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java [hunk 5]
```diff
@@ -161,7 +164,11 @@ public OperatorChain(
         Map<StreamEdge, RecordWriterOutput<?>> streamOutputMap =
                 new HashMap<>(outEdgesInOrder.size());
         this.streamOutputs = new RecordWriterOutput<?>[outEdgesInOrder.size()];
-        this.finishedOnRestore = finishedOnRestore;
+        this.finishedOnRestoreInput =
+                this.isFinishedOnRestore()
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java:168]: Very thanks for the explanation! if we used it in the constructor we indeed could not postpone the initialization into the subclass. I also agree with that we could leave it for a future improvement~

**Review by gaoyunhaii — APPROVED** [review]
Very thanks Piotr @pnowojski for updating the PR! LGTM and +1 to merge~

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Pull request merged**